Move to the bulk API.

webmat opened the pull request:
A quick test sending 1000 events with 3.0.5 vs the bulk API was obviously very
convincing. 3.0.5 took 1m12s while this version took 3s.

Missing:

- [x] Spec are only `allow`ing a call to be made to send_event (which as of #18  isn't happening anymore, but they don't fail). Must move this to `expect`.
- [x] Make the flush respect each event's potential custom `key` and `tag`
- [x] Make the buffering code easier to test
- [x] Add tests specific to the buffering.
- [x] Log warning and drop messages where the interpolated key field doesn't get set
- [x] Version bump to 3.1.0, this is a significant (but backwards compatible) change.

1st round of feedback

- [x] Dropped message only in debug, and pass event as a separate field
- [x] Replace the queueing in the plugin by `multi_receive` & appropriate instructions to use the Logstash persistent queue
- [x] Open an issue about replacing the retries with the DLQ: #30 
- [x] Repurpose `flush_size` to cap the size of API calls.

2nd round of feedback

- [x] Remove `flush_size` and split payload before we hit the API limit of 5Mb per API call
- [x] Drop events that go above 1Mb, log a warning
- [x] Add settings for max event size and max payload size & document these two size limits with appropriate links to the canonical doc.
- [x] Remove the `to_hash`. Try to keep with `Event` until it's time to convert to JSON.
- [x] Mention that batch size and max flush intervals are now handled at the Logstash pipeline level, for users who want to tweak them.
- [x] Review the changelog wrt latest developments

Discarded ideas:

- The Loggly API [does not return 413 entity too large](https://www.loggly.com/docs/api-overview/), when either one message or a full payload go over their respective size limit. Instead it returns 200 ok and doesn't ingest the event. So there's no need to check for 413.

Closes #18, closes #14, closes #22

webmat commented:
This was more work than I expected ;-)

PR #18 from 2016 did the *bulk* of the work, but didn't respect custom tag/keys. And I added a bunch of tests.

But I couldn't leave these improvements unmerged.

jsvd commented on docs/index.asciidoc:
```diff
@@ -63,8 +65,20 @@ output plugins.
+  * Default value is 100
+
+This plugin uses the bulk index api for improved indexing performance.
+To make efficient bulk api calls, we will buffer a certain number of
```
My concern with buffering data is that we lose control of acknowledgement for these events. All plugins tend to process data in a synchronous fashion as this is the only want to acknowledge that the events were processed correctly (or not).

If, for example, the persistent queue is enabled and this plugin accepts a few events, buffers them, then at some point tries to write to loggly but logstash is terminated for some reason, all buffered events are lost since, from the PQ's perspective, they were processed.

As an alternative I suggest moving to the same strategy as the elasticsearch output, that uses a multi event api `multi_receive(events)`: https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/71b2e530873728e60355e02adb3a9e7fa8c660e1/lib/logstash/outputs/elasticsearch/common.rb#L34-L37

jsvd commented on lib/logstash/outputs/loggly.rb:
```diff
@@ -88,31 +110,59 @@ class LogStash::Outputs::Loggly < LogStash::Outputs::Base
     key = event.sprintf(@key)
     tag = event.sprintf(@tag)
 
+    if expected_field = key[/%{(.*)}/, 1]
```
unless loggly throws an error when the key is shaped this way, I'd consider this a breaking change in behaviour as we'll start dropping data that we didn't before.

I see a few options here:
1. leave things as they were before and just log a warning
2. introduce a strict mode (off by default) and flip the switch in a next major version
3. add the option to [DLQ](https://www.elastic.co/guide/en/logstash/current/dead-letter-queues.html) these events (again off by default). This is a lot more work, may not be worth at this point.

webmat replied:
The API key is actually part of the Loggly URL. So instead of posting to URL `http://logs-01.loggly.com/bulk/API_KEY/tag/tag_name` we would be posting to an invalid URL:

`http://logs-01.loggly.com/bulk/%{key_from_field}/tag/tag_name`. So these events never made it to Loggly, they just caused invalid API calls.

Note that I'm not doing anything to the event, other outputs will receive it. I'm just not enqueuing it to ship to Loggly, to avoid the invalid call.

jsvd commented on lib/logstash/outputs/loggly.rb:
```diff
@@ -88,31 +110,59 @@ class LogStash::Outputs::Loggly < LogStash::Outputs::Base
 
+    if expected_field = key[/%{(.*)}/, 1]
+      @logger.warn "Skipping sending message to Loggly. No key provided (key='#{key}'). Make sure to set field '#{expected_field}'."
+      @logger.info "Dropped message: #{event.to_json}"
```
let's make this a debug level message, with big enough events this could flood the logs.
Also, we typically prefer structured logging instead of interpolation, so we can do this instead:

`@logger.info("Dropped message", :event => event.to_json)`

webmat replied:
Oh yes, both suggestions make total sense.

webmat commented:
Ok, moving to `multi_receive` gives the same benefits and will let me remove a bunch of code.

I ran this on the `Stud` implementation ([commit e58e82f](https://github.com/logstash-plugins/logstash-output-loggly/pull/29/commits/e58e82fcb291fb60a55b4fc244a6c37130f3af1f)) and on the `multi_receive` implementation.

```
time bin/logstash -e "input { generator { count => 1000 } }
  output {  loggly { key => \"$LOGGLY_KEY\"  } }"
```

Results:

- Stud: bin/logstash -e   84.30s user 2.76s system 353% cpu **24.601 total**
- multi_receive: bin/logstash -e   87.03s user 3.16s system 378% cpu **23.800 total**

I'll therefore remove the stud-related coded.

webmat commented:
@jsvd This is ready to be reviewed again & go out into the world.

Recap of what was done since your review:

- Removed dependency on Stud and support bulk API via `multi_receive` instead. Cleaned everything up wrt to the previous implementation.
- Retain the new attribute `flush_size` and use it to let users cap the size of their API calls. Default 100.
- Use structured logging in a few places that made sense & move the dropped event dump to `debug`.
- Clarify the plugin doc some more.

jsvd commented on CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+## 3.1.0
+  - Now use the Loggly bulk API. Adds settings `flush_size` and `idle_flush_time`.
```
idle_flush_time is no longer used

webmat replied:
Good catch.

jsvd commented on docs/index.asciidoc:
```diff
@@ -63,16 +64,28 @@ output plugins.
+===== `flush_size`
+
+  * Value type is <<number,number>>
+  * Default value is 100
```
since the default batch size in logstash is 125, I'd suggest either setting this default to that exact value or to fetch the configured `pipeline.batch.size` value. Otherwise we will always be splitting batches of 100/25 by default

webmat replied:
Good to know, I'll make this change.

jsvd commented on docs/index.asciidoc:
```diff
@@ -63,16 +64,28 @@ output plugins.
+
+This setting lets you ensure you don't send too many events at a time.
+See the bulk https://www.loggly.com/docs/http-bulk-endpoint/[API documentation]
+to estimate an appropriate amount of events, based on your expected event size.
```
should we reference that `You can send a maximum of 5MB per batch and up to 1MB per event.`?
Also, what happens when these limits are met? We'll probably have to add a new entry to the http status case [here](https://github.com/logstash-plugins/logstash-output-loggly/pull/29/files#diff-c561f06941a5987d4b5f9c26d87deb1bR200)  for `413 Request Entity Too Large`.
In the [es output we also have to deal with bulk size](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/b9b6910bae09f91ac92a2c72c968f81512cc6413/lib/logstash/outputs/elasticsearch/http_client.rb#L108-L128).
We may want to either split the batch into smaller ones on 413, retry exponentially forever to alert the user, or provide an (off by default) option to drop on 413

webmat replied:
In an ideal world, I would calculate the size of the payload as I build it (we're JSON-encoding objects one after another) and cut it off before 5Mb. But I figured this may be over-polishing for this pull request.

You are totally right that we do need to handle 413, though. Since it's one JSON object per line, I think it's straightforward enough to split the payload in two, then retry.

jsvd commented on lib/logstash/outputs/loggly.rb:
```diff
@@ -78,6 +83,12 @@ class LogStash::Outputs::Loggly < LogStash::Outputs::Base
+  # See the bulk API documentation to estimate an appropriate amount of events,
+  # based on your expected event size.
+  # https://www.loggly.com/docs/http-bulk-endpoint/
+  config :flush_size, :validate => :number, :default => 100, :required => true
```
if it makes things simpler I'd be totally ok with not providing this cap option and just let `pipeline.batch.size` and `pipeline.batch.interval` decide the event count (which the user can configure if necessary)

webmat replied:
I like the idea. Since we discussed splitting the payload on 413, might as well go straight to the approach of checking size as we encode, and remove this now unnecessary attribute.

I will still support splitting the payload on 413, in case their limits change from 5Mb/call.

jsvd commented on lib/logstash/outputs/loggly.rb:
```diff
@@ -88,31 +99,78 @@ class LogStash::Outputs::Loggly < LogStash::Outputs::Base
-    # Send event
-    send_event("#{@proto}://#{@host}/inputs/#{key}/tag/#{tag}", format_message(event))
-  end # def receive
+    [event.to_hash, key, tag]
```
I see here we're calling event.to_hash only to call `event.to_json` on `format_message`, so I think we can use proper events in the triple and let format_message convert directly from LogStash::Event to json

webmat replied:
I convert to hash here for testability, unfortunately. I have been hitting issues when comparing JSON objects that were equivalent, with content ordered differently.

But now that I spell it out, I guess a better approach would be to go to the length of decoding the JSON in the tests instead. The current approach is adding work in the main flow of the plugin for the purpose of tests, which is not ideal.

webmat commented:
This latest instalment satisfies all requirements discussed on Monday, except for one: I'm not testing for 413 response code yet. Submitting an event over their event size limit results in a 200 Ok, and the event never appears in Loggly, aka it fails silently. I haven't yet performed a test on the 5Mib limit for a whole payload, will do that first thing tomorrow.

Also of note, I think we should do a major version bump after all. There's lots of major changes here.

webmat commented:
@jsvd This PR is ready to review again. The Loggly API does not return 413 entity too large, so that's the only thing I skipped in your suggestions.

Approach is now make sure to flush an API call before reaching the API payload limit. There are two new params to override the Loggly size limits, and they are set to the current Loggly limits: `max_payload_size` and `max_event_size`.

I finally changed my mind on the version bump, and agree that this should be a big one. I went for v4.0.0.

jsvd commented:
any thoughts on the currently failing specs against 5.6 https://travis-ci.org/logstash-plugins/logstash-output-loggly/jobs/379210891 ?

webmat commented:
@jsvd Yes it's that JRuby 1.7 discrepancy. Fixed in https://github.com/logstash-plugins/logstash-output-loggly/pull/29/commits/80787572583a31f468788ae7b94b31069185d2a3

jsvd commented on lib/logstash/outputs/loggly.rb:
```diff
@@ -88,31 +105,114 @@ class LogStash::Outputs::Loggly < LogStash::Outputs::Base
-    # Send event
-    send_event("#{@proto}://#{@host}/inputs/#{key}/tag/#{tag}", format_message(event))
-  end # def receive
+    meta_event = {  key: key, tag: tag, event: event }
```
can we use 1.9 style hashes here? this guarantees it will work in logstash 5.6

webmat replied:
Yes, these were introduced in Ruby 1.9. Do you see an issue with using this style of hashes?

jsvd commented on spec/outputs/loggly_spec.rb:
```diff
@@ -24,42 +34,165 @@
+    it 'should send only one payload when everything fits' do
+      yielded_times = 0
+      output.build_message_bodies([event] * 10) do |body|
+        insist { body.lines.count } == 10
```
we're moving more and more away from insist, so for new tests I suggest using instead `expect(body.lines.count).to eq(10)`

webmat replied:
Makes sense. Less magic to have the `eq()` matcher. I'll change that.

webmat commented:
Just squashed to 2 commits (contributor and mine).

Shipping v4.0.0 as soon as CI is done :-)

webmat commented:
[v4.0.0](https://rubygems.org/gems/logstash-output-loggly/versions/4.0.0) is out

joshuabaird commented:
@webmat Thank you for your work on this!  When do new versions of the plugin usually hit the official releases/documentation here:  https://www.elastic.co/guide/en/logstash/current/plugins-outputs-loggly.html ?
